### ISSUE TITLE
`return x match {...}` crashes the compiler

### ISSUE THREAD

**Issue by robby-phd** [scala-js/scala-js#2928]
The following code somehow crashes the "jscode" phase (Scala 2.12.2, Scala.js 0.6.16):

```scala
class String(val value: Predef.String) extends AnyVal
class StringBug {
  def isTrue(s: String): Boolean = return s.value match {
    case "True" => true
    case _ => false
  }
}
```

Trace:

```
[trace] Stack trace suppressed: run last util-js/compile:compileIncremental for the full output.
[error] (util-js/compile:compileIncremental) java.lang.AssertionError: assertion failed: 
[error]   Found stats with case label def in non-match block at source-.../shared/src/main/scala/StringBug.scala,line-3,offset=107: {
[error]   case <synthetic> val x1: String = s;
[error]   case5(){
[error]     if ("True".==(x1))
[error]       matchEnd4(true)
[error]     else
[error]       case6()
[error]   };
[error]   case6(){
[error]     matchEnd4(false)
[error]   };
[error]   return matchEnd4(x: Boolean){
[error]     x
[error]   }
[error] }
[error]      while compiling: .../shared/src/main/scala/StringBug.scala
[error]         during phase: jscode
[error]      library version: version 2.12.2
[error]     compiler version: version 2.12.2
[error]   reconstructed args: -deprecation -feature -Xfatal-warnings -classpath .../js/target/scala-2.12/classes:.../shared/target/scala-2.12/classes:.../.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-js/scalajs-library_2.12/0.6.16/scalajs-library_2.12-0.6.16.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-js/scalajs-test-interface_2.12/0.6.16/scalajs-test-interface_2.12-0.6.16.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/com/lihaoyi/derive_2.12/0.4.4/derive_2.12-0.4.4.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/org/spire-math/jawn-parser_2.12/0.10.3/jawn-parser_2.12-0.10.3.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/com/lihaoyi/sourcecode_sjs0.6_2.12/0.1.3/sourcecode_sjs0.6_2.12-0.1.3.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.12.2/scala-reflect-2.12.2.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/com/lihaoyi/upickle_2.12/0.4.4/upickle_2.12-0.4.4.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/com/lihaoyi/utest_sjs0.6_2.12/0.4.5/utest_sjs0.6_2.12-0.4.5.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/com/lihaoyi/derive_sjs0.6_2.12/0.4.4/derive_sjs0.6_2.12-0.4.4.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/com/lihaoyi/sourcecode_2.12/0.1.3/sourcecode_2.12-0.1.3.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/com/lihaoyi/upickle_sjs0.6_2.12/0.4.4/upickle_sjs0.6_2.12-0.4.4.jar:.../.coursier/cache/v1/https/repo1.maven.org/maven2/com/github/ghik/silencer-lib_2.12/0.5/silencer-lib_2.12-0.5.jar -Xplugin:.../.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-js/scalajs-compiler_2.12.2/0.6.16/scalajs-compiler_2.12.2-0.6.16.jar -Xplugin:.../.coursier/cache/v1/https/repo1.maven.org/maven2/com/github/ghik/silencer-plugin_2.12/0.5/silencer-plugin_2.12-0.5.jar -unchecked -bootclasspath .../java/jre/lib/resources.jar:.../java/jre/lib/rt.jar:.../java/jre/lib/sunrsasign.jar:.../java/jre/lib/jsse.jar:.../java/jre/lib/jce.jar:.../java/jre/lib/charsets.jar:.../java/jre/lib/jfr.jar:.../java/jre/classes:.../.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.12.2/scala-library-2.12.2.jar
[error] 
[error]   last tree to typer: TypeTree(class String)
[error]        tree position: line 3 of .../shared/src/main/scala/StringBug.scala
[error]             tree tpe: String
[error]               symbol: final class String in package lang
[error]    symbol definition: final class String extends Serializable with Comparable with CharSequence (a ClassSymbol)
[error]       symbol package: java.lang
[error]        symbol owners: class String
[error]            call site: class StringBug in package <empty>
[error] 
[error] == Source file context for tree position ==
[error] 
[error]      1 class String(val value: Predef.String) extends AnyVal
[error]      2 class StringBug {
[error]      3   def isTrue(s: String): Boolean = return s.value match {
[error]      4     case "True" => true
[error]      5     case _ => false
[error]      6   }
[error]      7 }
```

Perhaps the cause is `` val x1: String``, which perhaps should be  ``val x1: Predef.String``.

The following works around the issue:

```scala
class String(val value: Predef.String) extends AnyVal
class StringBug {
  def isTrue(s: String): Boolean = s.value match { // remove return
    case "True" => true
    case _ => false
  }
}
```

**Comment by sjrd** [comment]
Thanks for the report.

**Comment by sjrd** [comment]
Minimized as follows:
```scala
class ReturnMatchBug {
  def isTrue(s: String): Boolean = return s match {
    case "True" => true
    case _      => false
  }
}
```
Nothing to with shadowing `String` nor even with value classes. It's the `return s match { ... }` that triggers it. After mixin, we get:
```scala
package helloworld {
  class ReturnMatchBug extends Object {
    def isTrue(s: String): Boolean = {
      case <synthetic> val x1: String = s;
      case5(){
        if ("True".==(x1))
          matchEnd4(true)
        else
          case6()
      };
      case6(){
        matchEnd4(false)
      };
      return matchEnd4(x: Boolean){
        x
      }
    };
    def <init>(): helloworld.ReturnMatchBug = {
      ReturnMatchBug.super.<init>();
      ()
    }
  }
}
```
Notice how the `return` found its way inside the `Block` and around the `matchEnd`. scalac, Why!?

**Comment by sjrd** [comment]
Fixed in 3544341c57a2d1aa4a978e3eac785e3686921f38
